CMS-Enterprise/easi-app
# [NOREF] Add script for calling deadcode

# NOREF

## Description

- Introduces a script for calling go's [deadcode tool](https://go.dev/blog/deadcode).
- Builds in a few features
  - Automatically installs the latest version of deadcode
  - Allows extra flags to be passed via the CLI (like `./scripts/go_deadcode -test`)
  - Maintains a regex-compatible list of exclusions for customizing the script to our codebase.

> [!NOTE]
> I am not intending (at this time) to introduce this as part of our `pre-commit` or CI/CD checks -- this is more of an informational script to help identify areas of potential cleanup, and is currently just meant to be run manually!

## How to test this change

- Run `./scripts/go_deadcode.sh`, and you'll see a list of unreachable/dead code
- Run `./scripts/go_deadcode.sh -test`, and you'll see a _smaller_ list of unreachable code (adding `-test` also checks for usage in test files)

> [!IMPORTANT]
> _Both_ of these commands (with and without `-test`) can be useful.
> - Calling _without_ `-test` is useful for finding functions that might only ever be used in test files (and therefore might be candidates for removal)
> - Calling _with_ `-test` is useful for hiding things like helper methods that are only ever used in tests (though, consider moving them to pkg/testhelpers so they get ignored automatically by the `exclusions` in the script!)

## PR Author Checklist




## PR Reviewer Guidelines

- It's best to pull the branch locally and test it, rather than just looking at the code online!
- When approving a PR, provide a reason _why_ you're approving it
  - e.g. "Approving because I tested it locally and all functionality works as expected"
  - e.g. "Approving because the change is simple and matches the Figma design"
- Don't be afraid to leave comments or ask questions, especially if you don't understand why something was done! (This is often a great time to suggest code comments or documentation updates)
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
